### PR TITLE
Dikin, Vaidya, John walk

### PR DESCRIPTION
**Pull request by AlexManochis**
This PR includes c++ and R wrappers for the three c++ implementations of vaidya walk, dikin walk and john walk in https://github.com/rzrsk/vaidya-walk.

1) I have modified the original c++ code to avoid the use of virtual classes. 
2) I have used the volesti c++ code structure to use the code from (1).
3) I have implemented R wrappers in `sample_points.cpp` in `R-proj/src` folder.
4) I have updated Rd files.

### CODE REVIEW

File: R-proj/src/sample_points.cpp [hunk 1]
```diff
@@ -22,7 +24,8 @@
 template <typename Polytope, typename RNGType, typename PointList, typename NT, typename Point>
 void sample_from_polytope(Polytope &P, RNGType &rng, PointList &randPoints, unsigned int const& walkL, unsigned int const& numpoints,
         bool const& gaussian, NT const& a, NT const& L, bool const& boundary, Point const& StartingPoint, unsigned int const& nburns,
-        bool const& set_L, bool const& cdhr, bool const& rdhr, bool const& billiard, bool const& ball_walk)
+        bool const& set_L, bool const& cdhr, bool const& rdhr, bool const& billiard, bool const& ball_walk, bool const& dikin,
```
Comment (vissarion) [inline R-proj/src/sample_points.cpp]: please use at least an `enum` this list of `bool`s is becoming too long

File: include/random_walks/ellipsoid_walks/dikin_walker.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+class DikinWalker {
+public:
+
+    DikinWalker() {}
+
+    //DikinWalker(const Eigen::Matrix<Dtype, Eigen::Dynamic, 1>& initialization, const Eigen::Matrix<Dtype, Eigen::Dynamic,
+    //        Eigen::Dynamic>& cons_A, const Eigen::Matrix<Dtype, Eigen::Dynamic, 1>& cons_b, const Dtype r) :
+    //        Walker<Dtype>(initialization, cons_A, cons_b), r_(r){}
+
+    void init(const Eigen::Matrix<Dtype, Eigen::Dynamic, 1> &initialization, const Eigen::Matrix <Dtype,
```
Comment (vissarion) [inline include/random_walks/ellipsoid_walks/dikin_walker.h]: why not at construction time ?

File: include/random_walks/ellipsoid_walks/john_walker.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+//namespace pwalk {
+
+template <typename Dtype>
+class JohnWalker {
+public:
+
+    JohnWalker() {}
+
+    //JohnWalker(const Eigen::Matrix<Dtype, Eigen::Dynamic, 1> &initialization,
```
Comment (vissarion) [inline include/random_walks/ellipsoid_walks/john_walker.h]: similar as above

File: include/random_walks/ellipsoid_walks/math_functions.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// Code from https://github.com/rzrsk/vaidya-walk
+
+#ifndef PWALK_UTIL_MATH_FUNCTIONS_HPP_
+#define PWALK_UTIL_MATH_FUNCTIONS_HPP_
+
+#include <Eigen/Dense>
+#include <boost/random.hpp>
+#include <cmath>
+
+//namespace pwalk {
```
Comment (vissarion) [inline include/random_walks/ellipsoid_walks/math_functions.h]: if not used remove it

File: include/random_walks/ellipsoid_walks/math_functions.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    static rng_t gen(1234567);
+    static boost::uniform_real<Dtype> random_distribution(a, b);
+    static boost::variate_generator<rng_t&, boost::uniform_real<Dtype> >
+    variate_generator(gen, random_distribution);
+
+    return variate_generator();
+}
+
+template <typename Dtype>
+Dtype gaussian_density(const Eigen::Matrix<Dtype, Eigen::Dynamic, 1>& x, const Eigen::Matrix<Dtype, Eigen::Dynamic, 1>& mu, const Eigen::Matrix<Dtype, Eigen::Dynamic, Eigen::Dynamic>& sqrt_cov) {
```
Comment (vissarion) [inline include/random_walks/ellipsoid_walks/math_functions.h]: too long line

File: include/random_walks/ellipsoid_walks/john_walker.h [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        return curr_sample_;
+    }
+
+    void proposal(Eigen::Matrix<Dtype, Eigen::Dynamic, 1> &new_sample) {
+        Eigen::Matrix<Dtype, Eigen::Dynamic, 1> gaussian_step = Eigen::Matrix<Dtype, Eigen::Dynamic, 1>::Zero(
+                this->nb_dim_);
+        sample_gaussian<Dtype>(this->nb_dim_, 0., 1., gaussian_step);
+
+        // get hessian
+        Eigen::Matrix <Dtype, Eigen::Dynamic, Eigen::Dynamic> new_sqrt_inv_hess = Eigen::Matrix<Dtype, Eigen::Dynamic, Eigen::Dynamic>::Zero(
```
Comment (vissarion) [inline include/random_walks/ellipsoid_walks/john_walker.h]: please wrap long lines

**Review by vissarion — COMMENTED** [review]
Thanks Alex for your PR! I have a few comments.

File: include/random_walks/ellipsoid_walks/john_walker.h (hunk 6 above)
Comment (AlexManochis) [reply include/random_walks/ellipsoid_walks/john_walker.h]: Thanks. I wrapped all long lines.

File: include/random_walks/ellipsoid_walks/math_functions.h (hunk 5 above)
Comment (AlexManochis) [reply include/random_walks/ellipsoid_walks/math_functions.h]: I wrapped it among others.

File: include/random_walks/ellipsoid_walks/math_functions.h (hunk 4 above)
Comment (AlexManochis) [reply include/random_walks/ellipsoid_walks/math_functions.h]: Thanks. I have removed all unused comments

File: R-proj/src/sample_points.cpp (hunk 1 above)
Comment (AlexManochis) [reply R-proj/src/sample_points.cpp]: Thanks. I have implemented it.

File: include/random_walks/ellipsoid_walks/dikin_walker.h (hunk 2 above)
Comment (AlexManochis) [reply include/random_walks/ellipsoid_walks/dikin_walker.h]: Thanks. I have removed the init function.

File: include/random_walks/ellipsoid_walks/john_walker.h (hunk 3 above)
Comment (AlexManochis) [reply include/random_walks/ellipsoid_walks/john_walker.h]: done as above

**Review by TolisChal — APPROVED** [review]

**Pull request merged**